C Device API (#31)
apache/tvm-rfcs: rfcs/0031-devices-api.md

# Motivation
[motivation]: #motivation

When using an accelerator, such as the [Arm&reg; Ethos&trade;-U](https://github.com/apache/tvm-rfcs/pull/11), an Embedded Real-Time Operating System (RTOS) will provide a device abstraction to access the device resource. When using these abstractions, TVM needs to understand how to interact with a device for a given platform.

Taking the common example of a UART interface (imagining the accelerator is communicated to via this interface); in Zephyr, this would look similar to:

```c
#include <zephyr.h>
#include <device.h>

struct device *uart_dev = device_get_binding("USART0");

char data[] = "Hello World!\r\n";
uart_tx(uart_dev, data, sizeof(data), 100);
```

Whereas in CMSIS, this would look more similar to:

```c
ARM_DRIVER_USART* uart_dev = &Driver_USART0;
uart_dev->Initialize(NULL);

char data[] = "Hello World!\r\n";
uart_dev->Send(data, sizeof(data)/sizeof(data[0]));
```

In this example, you can see the diversity of RTOS implementations for drivers and why it's required to provide a flexible abstraction to pass devices for micro targets.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## User App

> `tvm_device_<device>_t`s are implemented for each RTOS or platform required, these are included by the user who chooses as appropriate for their application. Notably, to avoid dynamic allocation, the user must provide the `tvm_device_<device>_t` struct and initialise it rather than it being created and setup for them in the API. This is augmented by named functions for each device, examples in the case of the "woofles" accelerator:

Reviewer: clearer would be "TVM presumes that the RTOS, platform, or user application defines a struct type `tvm_device_<device>_t`." Also state in the last sentence here that the user must implement a set of named functions.

Author: I've tweaked it a bit :smile_cat:

> ```c

Reviewer: can we include my life cycle picture and annotate with where each function is called? also can we document exactly why TVM calls each function and what it's expected to do (e.g. pre- and post-conditions)?

Author: I think this is included further down the RFC in the outline of an executor and explored further in future work? The hooks are intentionally kept simply named so they can be interpreted for a device, from my experience the `TVMDeviceWooflesActivate` could span from setting a pin high to sending a series of commands over a communications bus? The main thing I'm trying to communicate is that `Activate` is called before `Open` which is documented in the executor flow, this might not be the best way of demonstrating it though so please feel free to suggest an alternative.

Reviewer: i think the reason i'm hoping for a life-cycle diagram is that the actions Activate/Open could take are quite varied. i agree the call scheme below is laid out in call order, but there's nothing explicit which states this so you just have to know that in advance. could you add something more explicit? i think the life-cycle diagram is the best thing as we can iterate on it as we add features here.

typedef void* tvm_device_woofles_t; // Called by User App
int32_t TVMDeviceWooflesInit(tvm_device_woofles_t* tvm_dev, ...); // Called by User App
int32_t TVMDeviceWooflesActivate(tvm_device_woofles_t* tvm_dev); // Called by generated code
int32_t TVMDeviceWooflesOpen(tvm_device_woofles_t* tvm_dev); // Called by generated code
int32_t TVMDeviceWooflesClose(tvm_device_woofles_t* tvm_dev); // Called by generated code
int32_t TVMDeviceWooflesDeactivate(tvm_device_woofles_t* tvm_dev); // Called by generated code
int32_t TVMDeviceWooflesDestroy(tvm_device_woofles_t* tvm_dev); // Called by User App
```

Which is implemented as part of a User App:
```c
#include <tvm/runtime/device.h>
#include <tvm/device/woofles/zephyr.h>

struct device* woofles_zephyr_device = device_get_binding("WOOFLES0");
tvm_device_woofles_t accelerator; // Opaque type for accelerator device
TVMDeviceWooflesInit(&accelerator, woofles_zephyr_device);

struct tvmgen_mynetwork_devices devices {
    .accelerator = accelerator
};

int32_t ret = tvmgen_mynetwork_run(
    ...,
    &devices
);

TVMDeviceDestroy(&accelerator);
```

## Platform Structures

> Users can take a implementations from `src/runtime/crt/device` and headers from `include/runtime/crt/device` which maps to their platform device implementation. In the case of a bare metal environment, this would default to a void pointer as there's no information available.

Reviewer: ```suggestion
Users can take a implementations from `src/runtime/crt/device` and headers from `include/runtime/crt/device` which maps to their platform device implementation. The simplest definition of `tvm_device_<device>_t` is `void*` as no information is provided to TVM.
```

```c
typedef tvm_device_woofles_t void*;
```

For RTOS implementations, a structure can be created such as this simple Zephyr wrapper (include/runtime/crt/platform/zephyr.h):

```c
#include <device.h>

typedef struct {
    struct device* dev;
} tvm_device_woofles_t;
```

> This enables the OS maximum control over the resources required and provides the opportunity to craft code in whichever way is most idiomatic for that platform, such as if an additional locking mechanism is required:

Reviewer: ```suggestion
This enables the OS maximum control over the resources required, allows the user application to consolidate the memory used by the device control structures, and provides the opportunity to craft code in whichever way is most idiomatic for that platform, such as if an additional locking mechanism is required:
```

int32_t TVMDeviceWooflesDeactivate(tvm_device_woofles_t* tvm_dev); // Called by generated code
```

These can all be implemented using the user-opaque context `tvm_device_<device>_t*`, enabling the majority of TVM code to be portable between RTOS implementations; importantly this applies to those called within operator functions (see below). The executors are agnostic to the underlying device implementation and simply get passed the relevant device pointer which is then passed to the correct symbol.

## Platform Device API
To allow setting of platform specifics into the opaque struct, these should be defined in the platform header. Alongside the header, an additional file will provide implementations (`src/runtime/crt/device/<device>/<platform>.c`):
```c
int32_t TVMDeviceWooflesInit(tvm_device_t* tvm_dev, struct device* zephyr_dev) {
    tvm_dev->device = zephyr_dev;
}
```
This simple wrapper enables type checking of these functions and defining a clear translation boundary between the underlying OS implementation and TVM.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## Entrypoint

> The entrypoint API defined in [Embedded C Runtime Interface](https://discuss.tvm.apache.org/t/rfc-utvm-embedded-c-runtime-interface/9951) is augmented with the `devices` structure which contains implemented `tvm_device_t` `struct`s for each device used by the network:

Reviewer: ```suggestion
The entrypoint API defined in [Embedded C Runtime Interface](https://discuss.tvm.apache.org/t/rfc-utvm-embedded-c-runtime-interface/9951) is augmented with the `tvm_mynetwork_devices` structure which contains implemented `tvm_device_t` `struct`s for each device used by the network:
```

```
typedef struct {
    struct tvm_device_woofles_t* woofles
} tvmgen_mynetwork_devices;
```

These are re-cast to `void *` when entering the AOT main function to pass it without TIR understanding the struct types.

```c
int32_t tvmgen_mynetwork_run(
    ...,
    struct tvmgen_mynetwork_devices* devices
) {
    tvmgen_mynetwork_run_model(
        ...,

>         devices->host,

Reviewer: shall we just pass the devices struct and use `tir.tvm_struct_get`?

        devices->accelerator
    );
}
```

## Executor Function
Each operator is provided with a single device object which can be abstracted and passed as the `void* resource_handle`. The main function calls into the device API to setup and teardown resources before and after each operator call.

```c
int32_t tvmgen_mynetwork_run_model(..., device0, device1) {

>     TVMDeviceWooflesActivate(device0); // Could reserve or enable certain circuitry ahead of time

Reviewer: would be great to say something about the memory and perhaps include a note that memory copies into the device API-managed memory may occur before TVMDeviceWooflesOpen

Author: I added some words to that effect, essentially stating that the memory flow is outside this RFC as it stands right now but we should be aware that means it could happen at any time.

int32_t TVMDeviceWooflesOpen(tvm_device_woofles_t* tvm_dev) {
    k_mutex_lock(&tvm_dev->lock, K_FOREVER);
}

int32_t TVMDeviceWooflesClose(tvm_device_woofles_t* tvm_dev) {
    k_mutex_unlock(&tvm_dev->lock);
}

int32_t TVMDeviceWooflesDeactivate(tvm_device_woofles_t* tvm_dev) {}

int32_t TVMDeviceWooflesDestroy(tvm_device_woofles_t* tvm_dev) {
    tvm_dev->dev = NULL;
}
```

Whereas for CMSIS, you can use the platform-specific function to encapsulate the API to our imaginary UART accessed accelerator:

```c
typedef struct {
    ARM_DRIVER_USART* dev;
} tvm_device_uart_accel_t;

int32_t TVMDeviceUartAccelInit(tvm_device_uart_accel_t* tvm_dev, ARM_DRIVER_USART* uart_dev) {
    uart_dev->Initialize(NULL);
    tvm_dev->dev = uart_dev;
}

int32_t TVMDeviceUartAccelActivate(tvm_device_uart_accel_t* tvm_dev) {}
int32_t TVMDeviceUartAccelOpen(tvm_device_uart_accel_t* tvm_dev) {}
int32_t TVMDeviceUartAccelClose(tvm_device_uart_accel_t* tvm_dev) {}
int32_t TVMDeviceUartAccelDeactivate(tvm_device_uart_accel_t* tvm_dev) {}

int32_t TVMDeviceUartAccelDestroy(tvm_device_uart_accel_t* tvm_dev) {
    tvm_dev->dev->Uninitialize();
}
```

## Operator Usage

> Each operator would be expected to utilise one device structure and be passed that as the `resource_handle` parameter, making the assumption that each operator or variant of an operator is only bound to one device at a time. In the following example it can be seen how a accelerators interface is implemented per platform to take this void pointer and call the platform specific driver code.

Reviewer: could you specify how TVM knows which device structure an operator uses? also, can you qualify this is just for the C runtime and c++ runtime when BYOC targets are linked in?

Author: I'd rather avoid using the BYOC branding and sticking to the technical terms given BYOC is fairly heavily overloaded, I've added a few more words here to illustrate the `Target` configuration of this.

```c
// Operator takes opaque resource_handle
int32_t my_operator(..., void* resource_handle) {
    if (TVMDeviceWooflesInvoke(resource_handle, ...ins,outs,params...) != 0) {
        return -1;
    }
}

// Platform implementation
int32_t TVMDeviceWooflesInvoke(tvm_device_woofles_t* tvm_dev) {
    struct device* zephyr_dev = tvm_dev->dev;
    my_accelerator_invoke(
        zephyr_dev,
        ...ins,outs,params...
    );
}
```

## PrimFunc Resource Handle

> A `tir::Var` is added to `PrimFunc` in `include/tvm/tir/function.h` which enables a `PrimFunc` to track and use the `resource_handle` parameter. This will be used by both unpacked and packed APIs to pass the resource down without packing into `TVMValue`, instead as a `void *`. 

Reviewer: it kind of seems like we could get away without doing this by tracking which Target a function is going to run on and looking up the associated device API function in AOT/graph codegen. what do you think?

Author: We'd still need to pass this into the function call though to the operator though?

* Extends the C Interface API to add support for devices
* Resource management using `open`/`close` and `init`/`destroy` alongside opaque handles is a common pattern in C libraries

# Unresolved questions
[unresolved-questions]: #unresolved-questions

# Future possibilities
[future-possibilities]: #future-possibilities
This RFC aims to put in place the foundation of the Device API to start abstracting the various RTOS drivers. There are other flows that have been considered as extensions to this.

## Memory Copies
Movement of memory between additional devices which may be unable to communicate directly, this could take the form of simply:

```
// Copy from/to
int32_t TVMDeviceWooflesCopyFrom(tvm_device_t* source, void* destination);
int32_t TVMDeviceUartAccelCopyTo(void* source, tvm_device_t* destination);
```

And be integrated into the flow as follows:

```
TVMDeviceWooflesOpen(device1);
operator(..., device1) {
    // some work where device1 can read from memory directly
    // then the result is copied back
    TVMDeviceWooflesCopyFrom(device1, &buffer);
}
TVMDeviceWooflesClose(device1);

TVMDeviceUartAccelOpen(device2);
operator(..., device2) 

>     TVMDeviceUartAccelCopyTo(&buffer, device2);{

Reviewer: what if memory is needed across operator calls? does the device have to be Open'd for e.g. the whole runtime of the model when e.g. a parameter is pre-loaded and the device call is the last layer in the model?

Author: This is an example API for memory management in future possibilities, there's a much larger question as to how devices can specify all of the above as I'm unaware of which devices you're referring to and the specific characteristics in tandem with the model you're using.

[guide-level-explanation]: #guide-level-explanation

## User App

> TVM presumes that the RTOS, platform, or user application defines a struct type `tvm_device_<device>_t`, these are included by the user who chooses an implementation as appropriate for their application. Notably, to avoid dynamic allocation, the user must provide the `tvm_device_<device>_t` struct and initialise it rather than it being created and setup for them in the API. TVM then expects an implementation of the named functions for each device, examples in the case of the "woofles" accelerator:

Reviewer: could you also explicitly highlight where `<device>` comes from from a user PoV?

Author: Is this not covered by the example just below this block?

Author: Nope, that was not the question, amended the text to clarify the origin.

[reference-level-explanation]: #reference-level-explanation

## Entrypoint

> The entrypoint API defined in [Embedded C Runtime Interface](https://discuss.tvm.apache.org/t/rfc-utvm-embedded-c-runtime-interface/9951) is augmented with the `tvm_mynetwork_devices` structure which contains implemented `tvm_device_t` `struct`s for each device used by the network:

Reviewer: i forget if we discussed this already, but @manupa-arm 's RFC used a different casing so let's just quickly address this here: google c++ styleguide would have us call this `TvmGenMynetworkDevices`. I kind of like the `_` separator however, since this name is generated.

Author: I've intentionally left this as the C++ style as that's the current standard in TVM, @areusch suggested changing it [would likely require an RFC in itself and updating the style guide](https://github.com/apache/tvm/pull/8280#issuecomment-888554759).

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> Is coupling `Target`s with this Device API to the C Runtime acceptable? From the authors point of view this seems an acceptable trade off given the devices won't function correctly without the flow control.

Reviewer: i think it's likely that [TargetDevice](https://github.com/apache/tvm/pull/8892) will play a role here.

Author: Is there a `TargetDevice` RFC I can reference? Otherwise I'd assume `TargetDevice` is speculative rather than something we should include here?

Reviewer: sorry not yet :/. it's sort of teased in the future work section of my [Artifact pre-RFC](https://discuss.tvm.apache.org/t/introduce-artifact-a-container-for-generated-code/11099). i'm more including this pointer here as an "early FYI" how i'd like to solve this (e.g. using the TargetDevice name as the mechanism to bind a sub-target to a Device API impl).
